File: app/src/main/java/ru/michanic/mymot/UI/Adapters/SectionItemsListAdapter.java
package ru.michanic.mymot.UI.Adapters;

import android.content.ClipData;
import android.content.Context;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.BaseAdapter;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import java.util.List;

import de.halfbit.pinnedsection.PinnedSectionListView;
import ru.michanic.mymot.Models.Category;
import ru.michanic.mymot.Models.Model;
import ru.michanic.mymot.Models.SectionModelItem;
import ru.michanic.mymot.MyMotApplication;
import ru.michanic.mymot.Protocols.Const;
import ru.michanic.mymot.R;
import ru.michanic.mymot.UI.Cells.CatalogSliderCell;

public class SectionItemsListAdapter extends BaseAdapter implements PinnedSectionListView.PinnedSectionListAdapter {

    List<SectionModelItem> items;

    public SectionItemsListAdapter(List<SectionModelItem> items) {
        this.items = items;
    }

    public void setItems(List<SectionModelItem> items) {
        this.items = items;
    }

    @Override
    public boolean isItemViewTypePinned(int viewType) {
        return viewType == SectionModelItem.SECTION_TITLE;
    }

    @Override
    public int getCount() {
        return items.size();
    }

    @Override
    public Object getItem(int position) {
        return items.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override public int getViewTypeCount() {
        return 4;
    }

    @Override public int getItemViewType(int position) {
        return items.get(position).type;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        final SectionModelItem item = items.get(position);
        View view = convertView;

        switch (item.type) {
            case SectionModelItem.SECTION_TITLE:
                view = View.inflate(parent.getContext(), R.layout.cell_section_title, null);
                TextView title = (TextView) view.findViewById(R.id.section_title);
                title.setText(item.getSectionTitle());
                break;

            case SectionModelItem.LIST_MODEL:
                view = View.inflate(parent.getContext(), R.layout.cell_models_list, null);
                ImageView imageView = (ImageView) view.findViewById(R.id.cell_image);
                TextView modelTitle = (TextView) view.findViewById(R.id.model_title);
                TextView years = (TextView) view.findViewById(R.id.model_years);

                Model model = item.getModel();
                Picasso.get().load(Const.DOMAIN + model.getPreview_picture()).placeholder(R.drawable.ic_placeholder).into(imageView);
                modelTitle.setText(model.getName());
                years.setText(model.getYears());
                break;

            case SectionModelItem.SIMPLE_CELL:
                view = View.inflate(parent.getContext(), R.layout.cell_simple, null);
                TextView simpleTitle = (TextView) view.findViewById(R.id.textView);
                simpleTitle.setText(item.getPropertyTitle());
                break;

            case SectionModelItem.PRICE_CELL:
                view = View.inflate(parent.getContext(), R.layout.cell_price, null);
                TextView priceTitle = (TextView) view.findViewById(R.id.textView);
                EditText priceValue = (EditText) view.findViewById(R.id.priceValue);

                final String propertyTitle = item.getPropertyTitle();

                priceValue.addTextChangedListener(new TextWatcher() {
                    @Override
                    public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                    }

                    @Override
                    public void onTextChanged(CharSequence s, int start, int before, int count) {
                        if (propertyTitle == SectionModelItem.PRICE_FROM_NAME) {
                            try {
                                int priceFrom = Integer.parseInt(s.toString());
                                MyMotApplication.searchManager.setPriceFrom(priceFrom);
                            } catch(NumberFormatException nfe) { }

                        } else if (propertyTitle == SectionModelItem.PRICE_FOR_NAME) {
                            try {
                                int priceFor = Integer.parseInt(s.toString());
                                MyMotApplication.searchManager.setPriceFor(priceFor);
                            } catch(NumberFormatException nfe) { }
                        }
                        item.setPropertyValue(s.toString());
                    }

                    @Override
                    public void afterTextChanged(Editable s) {
                    }
                });

                priceTitle.setText(propertyTitle);
                priceValue.setText(item.getPropertyValue());
                break;
        }

        return view;
    }


}

File: app/src/main/java/ru/michanic/mymot/Kotlin/UI/Adapters/SectionItemsListAdapter.kt
package ru.michanic.mymot.Kotlin.UI.Adapters

import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import com.squareup.picasso.Picasso
import de.halfbit.pinnedsection.PinnedSectionListView.PinnedSectionListAdapter
import ru.michanic.mymot.Kotlin.Extensions.Font
import ru.michanic.mymot.Kotlin.Models.SectionModelItem
import ru.michanic.mymot.Kotlin.MyMotApplication
import ru.michanic.mymot.Kotlin.Protocols.Const
import ru.michanic.mymot.R

class SectionItemsListAdapter : BaseAdapter, PinnedSectionListAdapter {
   private var items: List<SectionModelItem>? = null

    constructor(nullableItems: List<SectionModelItem?>) {
        this.items = nullableItems.filterNotNull()
    }

    fun setItems(items: List<SectionModelItem?>) {
        this.items = items.filterNotNull()
    }


    override fun isItemViewTypePinned(viewType: Int): Boolean {
        return viewType == SectionModelItem.SECTION_TITLE
    }

    override fun getCount(): Int {
        return items!!.size  //TODO: ?
    }

    override fun getItem(position: Int): Any {
        return items!![position] //TODO: ?
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getViewTypeCount(): Int {
        return 4
    }

    override fun getItemViewType(position: Int): Int {
        return items!![position].type //TODO: ?
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View? {
        val item = items!![position] //TODO: ?
        var view = convertView
        when (item.type) {
            SectionModelItem.SECTION_TITLE -> {
                view = View.inflate(parent.context, R.layout.cell_section_title, null)
                val title = view.findViewById<View>(R.id.section_title) as TextView
                title.text = item.sectionTitle
                title.typeface = Font.oswald
            }
            SectionModelItem.LIST_MODEL -> {
                view = View.inflate(parent.context, R.layout.cell_models_list, null)
                val imageView = view.findViewById<View>(R.id.cell_image) as ImageView
                val modelTitle = view.findViewById<View>(R.id.model_title) as TextView
                val years = view.findViewById<View>(R.id.model_years) as TextView
                val volume = view.findViewById<View>(R.id.model_volume) as TextView
                val model = item.model
                Picasso.get().load(Const.DOMAIN + model?.preview_picture)
                    .placeholder(R.drawable.ic_placeholder).into(imageView)
                modelTitle.text = model?.name
                years.text = model?.years
                volume.text = model?.volume
            }
            SectionModelItem.SIMPLE_CELL -> {
                view = View.inflate(parent.context, R.layout.cell_simple, null)
                val simpleTitle = view.findViewById<View>(R.id.textView) as TextView
                simpleTitle.text = item.propertyTitle
            }
            SectionModelItem.PRICE_CELL -> {
                view = View.inflate(parent.context, R.layout.cell_price, null)
                val priceTitle = view.findViewById<View>(R.id.textView) as TextView
                val priceValue = view.findViewById<View>(R.id.priceValue) as EditText
                val propertyTitle = item.propertyTitle
                priceValue.addTextChangedListener(object : TextWatcher {
                    override fun beforeTextChanged(
                        s: CharSequence,
                        start: Int,
                        count: Int,
                        after: Int,
                    ) {
                    }

                    override fun onTextChanged(
                        s: CharSequence,
                        start: Int,
                        before: Int,
                        count: Int,
                    ) {
                        if (propertyTitle === SectionModelItem.PRICE_FROM_NAME) {
                            try {
                                val priceFrom = s.toString().toInt()
                                MyMotApplication.searchManager?.priceFrom = priceFrom
                            } catch (nfe: NumberFormatException) {
                                MyMotApplication.searchManager?.priceFrom = 0
                            }
                        } else if (propertyTitle === SectionModelItem.PRICE_FOR_NAME) {
                            try {
                                val priceFor = s.toString().toInt()
                                MyMotApplication.searchManager?.priceFor = priceFor
                            } catch (nfe: NumberFormatException) {
                                MyMotApplication.searchManager?.priceFor = 0
                            }
                        }
                        item.propertyValue = s.toString()
                    }

                    override fun afterTextChanged(s: Editable) {}
                })
                priceTitle.text = propertyTitle
                priceValue.setText(item.propertyValue)
            }
        }
        return view
    }

}